central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -523,7 +533,137 @@ func (ds *datastoreImpl) RemovePlopsByPod(ctx context.Context, id string) error
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+
+	return commandTag.RowsAffected()
+}
+
+// PruneOrphanedPLOPsByProcessIndicators prunes PLOPs that match process indicators without pods
+func (ds *datastoreImpl) PruneOrphanedPLOPsByProcessIndicators(ctx context.Context, pool postgres.DB, orphanWindow time.Duration) {
+	// TODO: Once it is guaranteed that all listening endpoints have PodUIDs, remove this function
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing deployments
+	query := fmt.Sprintf(deleteOrphanedPLOPDeploymentsAndPI, int(orphanWindow.Minutes()))
+	if _, err := pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by deployment: %v", err)
+	}
+
+	// Delete processes listening on ports orphaned because process indicators are orphaned due to
+	// missing pods.
+	query = fmt.Sprintf(deleteOrphanedPLOPPods, int(orphanWindow.Minutes()))
+	if _, err := pool.Exec(ctx, query); err != nil {
+		log.Errorf("failed to prune process listening on ports by pods: %v", err)
+	}
+}
+
+func (ds *datastoreImpl) readRowsToFindPLOPsWithNoProcessInformation(rows pgx.Rows) ([]string, error) {
+	var ids []string
+
+	for rows.Next() {
+		var serialized []byte
+
+		if err := rows.Scan(&serialized); err != nil {
+			return nil, pgutils.ErrNilIfNoRows(err)
+		}
+
+		var msg storage.ProcessListeningOnPortStorage
+		if err := proto.Unmarshal(serialized, &msg); err != nil {
+			return nil, err
+		}
+
+		process := msg.GetProcess()
+
+		if process == nil {
+			ids = append(ids, msg.GetId())
+		}
+	}
+	return ids, nil
+}
+
+func (ds *datastoreImpl) deletePLOPsInBatches(ctx context.Context, ids []string) (int64, error) {
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+	batchSize := 5000
+	numRecordsToDelete := len(ids)
+	for {
+
+		if len(ids) == 0 {
+			break
+		}
+
+		if len(ids) < batchSize {
+			batchSize = len(ids)
+		}
+
+		identifierBatch := ids[:batchSize]
+
+		if err := ds.storage.DeleteMany(ctx, identifierBatch); err != nil {
+			recordsDeleted := numRecordsToDelete - len(ids)
+			err = errors.Wrapf(err, "unable to delete the records.  Successfully deleted %d out of %d", recordsDeleted, numRecordsToDelete)
+			return int64(recordsDeleted), err
+		}
+
+		// Move the slice forward to start the next batch
+		ids = ids[batchSize:]
+	}
+	return int64(numRecordsToDelete), nil
+}
+
+func (ds *datastoreImpl) getPLOPsToDelete(ctx context.Context, pool postgres.DB) ([]string, error) {
+	ds.mutex.Lock()
+	defer ds.mutex.Unlock()
+
+	rows, err := pool.Query(ctx, getPotentiallyOrphanedPLOPs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer rows.Close()
+
+	plopsToDelete, err := ds.readRowsToFindPLOPsWithNoProcessInformation(rows)
+
+	return plopsToDelete, err
+}
+
+func (ds *datastoreImpl) RemovePLOPsWithoutProcessIndicatorOrProcessInfo(ctx context.Context, pool postgres.DB) (int64, error) {
```

Comment:
You could add `pool` to the `datastoreImpl` definition so we don't have to pass it around.

Reply:
Done